lib/readers/file-system.reader.ts

```diff
@@ -1,27 +1,52 @@
 import * as fs from 'fs';
 import * as path from 'path';
-import { Reader } from './reader';
+import { Reader, ReaderFileLackPersmissionsError } from './reader';
 
 export class FileSystemReader implements Reader {
   constructor(private readonly directory: string) {}
 
-  public list(): Promise<string[]> {
-    return fs.promises.readdir(this.directory);
+  public list(): string[] {
+    return fs.readdirSync(this.directory);
   }
 
-  public read(name: string): Promise<string> {
-    return fs.promises.readFile(path.join(this.directory, name), 'utf8');
+  public read(name: string): string {
+    return fs.readFileSync(path.join(this.directory, name), 'utf8');
   }
 
-  public async readAnyOf(filenames: string[]): Promise<string | undefined> {
-    try {
-      for (const file of filenames) {
-        return await this.read(file);
+  public readAnyOf(
+    filenames: string[],
+  ): string | undefined | ReaderFileLackPersmissionsError {
+    let firstFilePathFoundButWithInsufficientPermissions: string | undefined;
+
+    for (let idx=0; idx < filenames.length; idx++) {
```

Comment:
```suggestion
    for (let idx = 0; idx < filenames.length; idx++) {
```
nit: why idx not just id

Reply:
I'm used to call _index_ as _idx_ because _i_ is too short, and _id_ looks like 'unique id'